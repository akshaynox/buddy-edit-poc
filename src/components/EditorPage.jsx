import React, { useEffect, useState, useRef } from "react";
import ReactQuill, { Quill } from "react-quill";
import "quill/dist/quill.snow.css";
import io from "socket.io-client";
import { useParams, useLocation, useNavigate } from "react-router-dom";
import Delta from "quill-delta";
import toast from "react-hot-toast";
import UserClient from "../components/UserClient";
import QuillCursors from "quill-cursors";
import { FormattedMessage, useIntl } from "react-intl";
import generateRandomColor from "../utils/generateRandomColor";

Quill.register("modules/cursors", QuillCursors);

const EditorPage = () => {
  const { roomId } = useParams();
  const { search } = useLocation();
  const query = new URLSearchParams(search);
  const username = query.get("username");
  const reactNavigator = useNavigate();
  const intl = useIntl();

  const [socket, setSocket] = useState(null);
  const [documentState, setDocumentState] = useState(null); // New state to keep track of the document state
  const [clients, setClients] = useState([]); // New state for connected clients array

  const quillRef = useRef(null);

  // Define the modules
  const modules = {
    cursors: {
      hideDelayMs: 500,
      hideSpeedMs: 300,
      selectionChangeSource: null,
      transformOnTextChange: true,
    },
    toolbar: [
      [{ header: [1, 2, false] }],
      ["bold", "italic", "underline"],
      ["image", "code-block"],
    ],
  };

  useEffect(() => {
    //Quill initialization
    console.log(
      "1st time component rendering. So, Quill REF: " + quillRef.current
    );
    if (quillRef.current) return; //If already set Return i.e Object exists
    const editor = quillRef.current.getEditor();

    editor.setText("Loading...");
    //editor.setContents(staticDelta);  //For Testing
    quillRef.current = editor;
    console.log("1st useEffect For Quill Initialization");
  }, []);

  // -----------------------------------------2nd useEffect()    Starts --------------------------------------------------
  useEffect(() => {
    if (!quillRef.current) {
      console.error("Quill Editor is not initialized yet.");
      return;
    }
    console.log("2nd useEffect");

    const socketInstance = io("http://localhost:5000");
    setSocket(socketInstance);

    socketInstance.on("connect", () => {
      console.log("Socket connected, joining room with ID:", roomId);
      socketInstance.emit("join-room", { roomId, username });
    });

    // Event handlers for user join
    socketInstance.on("user-joined", (username) => {
      if (typeof socket === "undefined") return;
      setClients((prev) => [...prev, username]);

      const color = generateRandomColor();
      const cursors = quillRef.current.getEditor().getModule("cursors");
      const allCursors = cursors.cursors();
      console.log("All available cursors " + JSON.stringify(allCursors));
      if (!allCursors.some((cursor) => cursor.id === username)) {
        cursors.createCursor(username, username, color);
      }

      toast.success(`${username} ${intl.formatMessage({ id: "joined_room" })}`);
      socketInstance.emit("connected-users", clients); //Emit User Array
    });

    socketInstance.on("connected-users", handleConnectedUsers);

    socketInstance.on("initialize-document", (documentState) => {
      //console.log("Received document state:", documentState);
      const delta = new Delta(documentState);

      if (typeof documentState === "object" && documentState !== null) {
        quillRef.current.getEditor().setContents(delta, "silent");
        setDocumentState(delta);
      } else {
        console.error(
          "Document state is neither a string nor a valid object:",
          documentState
        );
        quillRef.current.getEditor().setText("Failed to load document.");
      }
      quillRef.current.getEditor().enable();
    });

    socketInstance.on("text-change", (data) => {
      if (data.username !== username) {
        const delta = new Delta(data.delta);
        quillRef.current.getEditor().updateContents(delta, "silent");
        //   console.log("ON text-change " + JSON.stringify(delta));
        setDocumentState((prevState) => {
          if (prevState) {
            const newDocumentState = prevState.compose(delta);
            // console.log("New document state after compose:", newDocumentState);
            return newDocumentState;
          } else {
            console.error("Previous state is undefined.");
            return new Delta(); // Reset to an empty Delta if undefined
          }
        });
      }
    });

    socketInstance.on("user-left", (username) => {
      setClients((prev) => prev.filter((user) => user !== username));

      const cursors = quillRef.current.getEditor().getModule("cursors");
      cursors.removeCursor(username, username);

      toast.success(`${username} ${intl.formatMessage({ id: "left_room" })}`);
      socketInstance.emit("connected-users", clients);
    });

    return () => {
      socketInstance.off("user-joined");
      socketInstance.off("connected-users");
      socketInstance.off("initialize-document");
      socketInstance.off("text-change");
      socketInstance.off("user-left");
      socketInstance.disconnect();
      if (quillRef.current)
        //object exists
        quillRef.current.getEditor().off("text-change");
      console.log("Cleaned up on component unmount");
    };
  }, [roomId, username]);

  // -----------------------------------------------------------2nd UseEffect ends--------------------------------------------------------------

  //Function to handle all connected clients
  const handleConnectedUsers = (users) => {
    console.log(`Updated list of connected users: ${users}`);
    setClients(users);
  };

  // Function to handle leaving the room
  const leaveRoom = () => {
    if (!socket) return;
    socket.emit("leave-room", { roomId, username });
    reactNavigator("/");
  };

  // --------------------------------3rd effect ----------------------------------------------
  useEffect(() => {
    if (!socket || !quillRef.current) return;

    const handleTextChange = (delta, oldDelta, source) => {
      if (source !== "user") return;

      let range = quillRef.current.getEditor().getSelection();
      if (range) {
        if (range.length == 0) {
          // User is actively typing, update the cursor position
          socket.emit("cursor-move", { roomId, username, cursorPos: range });
        } else {
          // User made a selection
          console.log("User has made a selection");
        }
      }

      socket.emit("text-change", { roomId, username, delta: delta });
      saveCurrentDocumentState();
    };

    const handleSelectionChange = (range) => {
      if (!socket) return;
      socket.emit("cursor-selection", { roomId, username, cursorPos: range });
      console.log("Selection-change of quill: " + JSON.stringify(range)); //{"index":0,"length":0}
    };

    socket.on("remote-cursor-selection", ({ username, cursorPos }) => {
      console.log(
        "Remote cursor selection point for " +
          username +
          " " +
          JSON.stringify(cursorPos)
      );
      const color = generateRandomColor();
      const cursors = quillRef.current.getEditor().getModule("cursors");
      const allCursors = cursors.cursors();
      console.log("All cursors " + JSON.stringify(allCursors));
      if (!allCursors.some((cursor) => cursor.id === username)) {
        cursors.createCursor(username, username, color);
      }

      cursors.moveCursor(username, cursorPos); // <== cursor data from previous step
      cursors.toggleFlag(username, true);
    });

    socket.on("remote-cursor-move", ({ username, cursorPos }) => {
      console.log(
        "Remote cursor move for " + username + " " + JSON.stringify(cursorPos)
      );
      const cursors = quillRef.current.getEditor().getModule("cursors");
      cursors.moveCursor(username, cursorPos); // <== cursor data from previous step
      cursors.toggleFlag(username, true);
    });

    quillRef.current.getEditor().on("text-change", handleTextChange);

    quillRef.current
      .getEditor()
      .on("selection-change", (range, oldRange, source) => {
        handleSelectionChange(range);
      });

    return () => {
      socket.off("remote-cursor-selection");
      socket.off("remote-cursor-move");

      if (quillRef.current) {
        quillRef.current.getEditor().off("text-change", handleTextChange);
        quillRef.current.getEditor().off("selection-change");
      }
    };
  }, [socket, roomId, username]);

  // Function to save the current document state to the server
  const saveCurrentDocumentState = () => {
    if (!quillRef.current) return;
    if (quillRef.current) {
      const currentContents = quillRef.current.getEditor().getContents(); // Get the current state of the editor
      const serializedContent = JSON.stringify(currentContents); // Serialize the state
      socket.emit("save-document", { roomId, content: serializedContent });
    }
  };

  // Function to copy room ID to clipboard
  const copyRoomId = () => {
    navigator.clipboard.writeText(roomId);
    toast.success(intl.formatMessage({ id: "room_id_clipboard" }));
  };

  return (
    <div className="mainWrap flex">
      <div className="aside flex-[0.2] bg-[#283236] p-4 text-white flex flex-col justify-center items-center gap-2">
        <div className="asideInner flex-1">
          <h3 className="text-xl font-bold">
            <FormattedMessage id="editing_users" />
          </h3>
          <div className="clientsList flex items-center flex-wrap gap-5 m-5">
            {clients.map((client, index) => (
              <UserClient
                key={index}
                username={client}
                isCurrentUser={client === username}
              />
            ))}
          </div>
        </div>
        <button
          className="btn copyBtn bg-white text-black w-[200px]"
          onClick={copyRoomId}
        >
          <FormattedMessage id="copy_room_id" />
        </button>
        <button
          className="btn leaveBtn bg-[#4aed88] text-black w-[200px]"
          onClick={leaveRoom}
        >
          <FormattedMessage id="leave" />
        </button>
      </div>
      <div className="editorWrap flex-1">
        <div className="editor-container" style={{ height: "100vh" }}>
          {" "}
          <ReactQuill
            ref={quillRef}
            theme="snow"
            modules={modules}
            placeholder={intl.formatMessage({ id: "start_collab" })}
            style={{ height: "100%" }}
          />
        </div>
      </div>
    </div>
  );
};

export default EditorPage;
