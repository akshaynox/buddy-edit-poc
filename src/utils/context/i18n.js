// ** React Imports
import { useState, createContext } from "react";

// ** Intl Provider Import
import { IntlProvider } from "react-intl";

// ** Core Language Data
import messagesEn from "../../assets/locales/en.json";
import messagesDe from "../../assets/locales/de.json";
import messagesFr from "../../assets/locales/fr.json";

// ** Menu msg obj
const menuMessages = {
  en: { ...messagesEn },
  de: { ...messagesDe },
  fr: { ...messagesFr },
};

// ** Create Context
const Context = createContext();
const DEFAULT_LANG = "en";

const IntlProviderWrapper = ({ children }) => {
  // ** States
  const [locale, setLocale] = useState(DEFAULT_LANG);
  const [messages, setMessages] = useState(menuMessages[DEFAULT_LANG]);

  // ** Switches Language
  const switchLanguage = (lang) => {
    setLocale(lang);
    setMessages(menuMessages[lang]);
  };

  return (
    <Context.Provider value={{ locale, switchLanguage }}>
      <IntlProvider
        key={locale}
        locale={locale}
        messages={messages}
        defaultLocale={DEFAULT_LANG}
      >
        {children}
      </IntlProvider>
    </Context.Provider>
  );
};

export { IntlProviderWrapper, Context as IntlContext };
