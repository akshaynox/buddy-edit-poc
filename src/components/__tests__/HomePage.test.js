import { cleanup, render, screen } from "@testing-library/react"
import HomePage from "../Homepage"
import { IntlProviderWrapper } from "../../utils/context/i18n"
import { BrowserRouter } from "react-router-dom";
import { renderWithReactIntl } from "./EditorPage.test";

afterEach(() => {
  cleanup();
})

test("should render homepage component", () => {
  render(
    <BrowserRouter>
      <IntlProviderWrapper>
        <HomePage />
      </IntlProviderWrapper>
    </BrowserRouter>
  );
  const homePageElement = screen.getByTestId("homepage");
  expect(homePageElement).toBeInTheDocument();
})

test("matches home page snapshot", async () => {
  const { asFragment } = renderWithReactIntl(
    <BrowserRouter>
      <IntlProviderWrapper>
        <HomePage />
      </IntlProviderWrapper>
    </BrowserRouter>
  );
  expect(asFragment()).toMatchSnapshot();
})