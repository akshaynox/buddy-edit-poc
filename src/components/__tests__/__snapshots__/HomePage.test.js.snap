// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches editor page snapshot 1`] = `
<DocumentFragment>
  <div
    class="mainWrap flex"
    data-testid="editorpage"
  >
    <div
      class="aside flex-[0.2] bg-[#283236] p-4 text-white flex flex-col justify-center items-center gap-2"
    >
      <div
        class="asideInner flex-1"
      >
        <h3
          class="text-xl font-bold"
        >
          Editing Users
        </h3>
        <div
          class="clientsList flex items-center flex-wrap gap-5 m-5"
        />
      </div>
      <button
        class="btn copyBtn bg-white text-black w-[200px]"
      >
        Copy Room ID
      </button>
      <button
        class="btn leaveBtn bg-[color:var(--btn-green)] text-black w-[200px]"
      >
        Leave
      </button>
    </div>
    <div
      class="editorWrap flex-1"
    >
      <div
        class="editor-container"
        style="height: 100vh;"
      >
         
        <div
          class="quill "
          style="height: 100%;"
        >
          <div
            class="ql-toolbar ql-snow"
          >
            <span
              class="ql-formats"
            >
              <span
                class="ql-header ql-picker"
              >
                <span
                  aria-controls="ql-picker-options-2"
                  aria-expanded="false"
                  class="ql-picker-label"
                  role="button"
                  tabindex="0"
                >
                  <svg
                    viewBox="0 0 18 18"
                  >
                     
                    <polygon
                      class="ql-stroke"
                      points="7 11 9 13 11 11 7 11"
                    />
                     
                    <polygon
                      class="ql-stroke"
                      points="7 7 9 5 11 7 7 7"
                    />
                     
                  </svg>
                </span>
                <span
                  aria-hidden="true"
                  class="ql-picker-options"
                  id="ql-picker-options-2"
                  tabindex="-1"
                >
                  <span
                    class="ql-picker-item"
                    data-value="1"
                    role="button"
                    tabindex="0"
                  />
                  <span
                    class="ql-picker-item"
                    data-value="2"
                    role="button"
                    tabindex="0"
                  />
                  <span
                    class="ql-picker-item ql-selected"
                    role="button"
                    tabindex="0"
                  />
                </span>
              </span>
              <select
                class="ql-header"
                style="display: none;"
              >
                <option
                  value="1"
                />
                <option
                  value="2"
                />
                <option
                  selected="selected"
                />
              </select>
            </span>
            <span
              class="ql-formats"
            >
              <button
                class="ql-bold"
                type="button"
              >
                <svg
                  viewBox="0 0 18 18"
                >
                   
                  <path
                    class="ql-stroke"
                    d="M5,4H9.5A2.5,2.5,0,0,1,12,6.5v0A2.5,2.5,0,0,1,9.5,9H5A0,0,0,0,1,5,9V4A0,0,0,0,1,5,4Z"
                  />
                   
                  <path
                    class="ql-stroke"
                    d="M5,9h5.5A2.5,2.5,0,0,1,13,11.5v0A2.5,2.5,0,0,1,10.5,14H5a0,0,0,0,1,0,0V9A0,0,0,0,1,5,9Z"
                  />
                   
                </svg>
              </button>
              <button
                class="ql-italic"
                type="button"
              >
                <svg
                  viewBox="0 0 18 18"
                >
                   
                  <line
                    class="ql-stroke"
                    x1="7"
                    x2="13"
                    y1="4"
                    y2="4"
                  />
                   
                  <line
                    class="ql-stroke"
                    x1="5"
                    x2="11"
                    y1="14"
                    y2="14"
                  />
                   
                  <line
                    class="ql-stroke"
                    x1="8"
                    x2="10"
                    y1="14"
                    y2="4"
                  />
                   
                </svg>
              </button>
              <button
                class="ql-underline"
                type="button"
              >
                <svg
                  viewBox="0 0 18 18"
                >
                   
                  <path
                    class="ql-stroke"
                    d="M5,3V9a4.012,4.012,0,0,0,4,4H9a4.012,4.012,0,0,0,4-4V3"
                  />
                   
                  <rect
                    class="ql-fill"
                    height="1"
                    rx="0.5"
                    ry="0.5"
                    width="12"
                    x="3"
                    y="15"
                  />
                   
                </svg>
              </button>
            </span>
            <span
              class="ql-formats"
            >
              <button
                class="ql-image"
                type="button"
              >
                <svg
                  viewBox="0 0 18 18"
                >
                   
                  <rect
                    class="ql-stroke"
                    height="10"
                    width="12"
                    x="3"
                    y="4"
                  />
                   
                  <circle
                    class="ql-fill"
                    cx="6"
                    cy="7"
                    r="1"
                  />
                   
                  <polyline
                    class="ql-even ql-fill"
                    points="5 12 5 11 7 9 8 10 11 7 13 9 13 12 5 12"
                  />
                   
                </svg>
              </button>
              <button
                class="ql-code-block"
                type="button"
              >
                <svg
                  viewBox="0 0 18 18"
                >
                   
                  <polyline
                    class="ql-even ql-stroke"
                    points="5 7 3 9 5 11"
                  />
                   
                  <polyline
                    class="ql-even ql-stroke"
                    points="13 7 15 9 13 11"
                  />
                   
                  <line
                    class="ql-stroke"
                    x1="10"
                    x2="8"
                    y1="5"
                    y2="13"
                  />
                   
                </svg>
              </button>
            </span>
          </div>
          <div
            class="ql-container ql-snow"
          >
            <div
              class="ql-editor ql-blank"
              contenteditable="true"
              data-gramm="false"
              data-placeholder="Start collaborating as a team..."
            >
              <p>
                <br />
              </p>
            </div>
            <div
              class="ql-clipboard"
              contenteditable="true"
              tabindex="-1"
            />
            <div
              class="ql-tooltip ql-hidden"
            >
              <a
                class="ql-preview"
                href="about:blank"
                rel="noopener noreferrer"
                target="_blank"
              />
              <input
                data-formula="e=mc^2"
                data-link="https://quilljs.com"
                data-video="Embed URL"
                type="text"
              />
              <a
                class="ql-action"
              />
              <a
                class="ql-remove"
              />
            </div>
            <div
              class="ql-cursors"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`matches home page snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-full m-0 bg-gradient-to-br from-[#005662] to-[#00334d]"
    data-testid="homepage"
    id="body"
  >
    <div
      class="fixed right-10 top-10"
    >
      <div
        data-testid="langswitcher"
      >
        <div>
          <select>
            <option
              selected=""
              value="en"
            >
              English
            </option>
            <option
              value="de"
            >
              German
            </option>
            <option
              value="fr"
            >
              French
            </option>
          </select>
        </div>
      </div>
    </div>
    <div
      class="homePageWrapper flex align items-center justify-center text-white h-screen"
    >
      <div
        class="formWrapper bg-[#283236] p-5 rounded-md w-[500px] max-w-[90%]"
      >
        <h4
          class="mainLabel mb-5 mt-0 text-xl font-bold"
        >
          Buddy Edit
        </h4>
        <div
          class="inputGroup flex flex-col"
        >
          <input
            class="inputBox"
            placeholder="Room Id"
            type="text"
            value=""
          />
          <input
            class="inputBox"
            placeholder="Username"
            type="text"
            value=""
          />
          <button
            class="btn joinBtn bg-[color:var(--btn-green)] w-25 ml-auto hover:bg-[#2b824c] text-black p-2 rounded-md min-w-4"
          >
            Join
          </button>
          <span
            class="createInfo mx-0 my-auto mt-5"
          >
            No invitation code? then create 
            <button
              class="createNewBtn text-[color:var(--btn-green)] border-b-[1px_solid_#4aed88] hover:cursor-pointer"
            >
              new room
            </button>
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
